app/controllers/subnets_controller.rb

```diff
@@ -40,21 +40,20 @@ def destroy
 
   # query our subnet dhcp proxy for an unused IP
   def freeip
-    not_found and return unless (s=params[:subnet_id].to_i) > 0
+    invalid_request and return unless (s=params[:subnet_id].to_i) > 0
     organization = params[:organization_id].blank? ? nil : Organization.find(params[:organization_id])
     location = params[:location_id].blank? ? nil : Location.find(params[:location_id])
     Taxonomy.as_taxonomy organization, location do
-      not_found and return unless (subnet = Subnet.authorized(:view_subnets).find(s))
+      render_403 and return unless (subnet = Subnet.authorized(:view_subnets).find(s))
       if (ip = subnet.unused_ip(params[:host_mac], params[:taken_ips]))
         render :json => {:ip => ip}
       else
-        # we don't want any failures if we failed to query our proxy
-        head :status => 200
+        not_found
       end
     end
   rescue => e
     logger.warn "Failed to query subnet #{s} for free ip: #{e}"
-    head :status => 500
+    generic_exception e
```

Comment:
generic_exception doesn't need to be called explicitly, it should be rescued (see app controller).  But there's also process_ajax_error in app controller, which is used in hosts controller, can that be reused?